[Python] Crash when referring to nondet_str

For the following program:
```python
foo = nondet_str()

if foo == "":
    foo = "foo"
```
ESBMC crashes:
```
$ esbmc ex.py --unwind 10
ESBMC version 7.11.0 64-bit x86_64 linux
Target: 64-bit little-endian x86_64-unknown-linux with esbmclibc
Parsing ex.py

Type checking warning:
ex.py:1: error: Name "nondet_str" is not defined  [name-defined]
Found 1 error in 1 file (checked 1 source file)

Converting
Generating GOTO Program
GOTO program creation time: 0.684s
GOTO program processing time: 0.003s
Starting Bounded Model Checking
Unwinding loop 11 iteration 1   file /home/ubuntu/git/esbmc/src/c2goto/library/string.c line 110 column 12 function strcmp
Unwinding loop 11 iteration 2   file /home/ubuntu/git/esbmc/src/c2goto/library/string.c line 110 column 12 function strcmp
Unwinding loop 11 iteration 3   file /home/ubuntu/git/esbmc/src/c2goto/library/string.c line 110 column 12 function strcmp
Unwinding loop 11 iteration 4   file /home/ubuntu/git/esbmc/src/c2goto/library/string.c line 110 column 12 function strcmp
Unwinding loop 11 iteration 5   file /home/ubuntu/git/esbmc/src/c2goto/library/string.c line 110 column 12 function strcmp
Unwinding loop 11 iteration 6   file /home/ubuntu/git/esbmc/src/c2goto/library/string.c line 110 column 12 function strcmp
Unwinding loop 11 iteration 7   file /home/ubuntu/git/esbmc/src/c2goto/library/string.c line 110 column 12 function strcmp
Unwinding loop 11 iteration 8   file /home/ubuntu/git/esbmc/src/c2goto/library/string.c line 110 column 12 function strcmp
Unwinding loop 11 iteration 9   file /home/ubuntu/git/esbmc/src/c2goto/library/string.c line 110 column 12 function strcmp
Not unwinding loop 11 iteration 10   file /home/ubuntu/git/esbmc/src/c2goto/library/string.c line 110 column 12 function strcmp
Symex completed in: 0.008s (279 assignments)
Caching time: 0.000s (removed 0 assertions)
Slicing time: 0.001s (removed 204 assignments)
Generated 41 VCC(s), 16 remaining after simplification (75 assignments)
No solver specified; defaulting to Boolector
Encoding remaining VCC(s) using bit-vector/floating-point arithmetic
ERROR: compute_pointer_offset, unexpected irep:
with
* source_value : symbol
    * name : nondet$symex::nondet0
    * renamelev : Level 1 (global)
    * level1_num : 0
    * level2_num : 0
    * thread_num : 0
    * node_num : 0
    * type : array
      * subtype : signedbv
          * width : 8
      * array_size : constant_int
          * value : 16
          * type : unsignedbv
            * width : 64
      * size_is_infinite : false
* update_field : constant_int
    * value : 15
    * type : unsignedbv
      * width : 64
* update_value : constant_int
    * value : 0
    * type : signedbv
      * width : 8
* type : array
  * subtype : signedbv
      * width : 8
  * array_size : constant_int
      * value : 16
      * type : unsignedbv
        * width : 64
  * size_is_infinite : false
Aborted (core dumped)
```
ESBMC version: 9381b01a5

--- a/regression/python/nondet_list17/test.desc
+++ b/regression/python/nondet_list17/test.desc
@@ -1,4 +1,4 @@
 CORE
 main.py
---unwind 16
+--unwind 17
 ^VERIFICATION SUCCESSFUL$

--- a/regression/python/nondet_list18/test.desc
+++ b/regression/python/nondet_list18/test.desc
@@ -1,4 +1,4 @@
 CORE
 main.py
---unwind 3 --nondet-str-length 3
+--unwind 4 --nondet-str-length 3
 ^VERIFICATION SUCCESSFUL$

--- a/regression/python/nondet_str10/main.py
+++ b/regression/python/nondet_str10/main.py
@@ -1,0 +1,5 @@
+s = nondet_str()
+original = s
+s = "modified"
+# This assertion should FAIL because s was reassigned
+assert s == original

--- a/regression/python/nondet_str10/test.desc
+++ b/regression/python/nondet_str10/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.py
+--incremental-bmc
+^VERIFICATION FAILED$

--- a/regression/python/nondet_str5/main.py
+++ b/regression/python/nondet_str5/main.py
@@ -1,0 +1,6 @@
+foo = nondet_str()
+
+if foo == "":
+    foo = "foo"
+
+assert foo == "foo" or foo == "" or foo != "foo"

--- a/regression/python/nondet_str5/test.desc
+++ b/regression/python/nondet_str5/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.py
+--incremental-bmc
+^VERIFICATION SUCCESSFUL$

--- a/regression/python/nondet_str6/main.py
+++ b/regression/python/nondet_str6/main.py
@@ -1,0 +1,6 @@
+foo = nondet_str()
+
+if foo == "":
+    foo = "foo"
+
+assert foo == "foo" or foo == ""

--- a/regression/python/nondet_str6/test.desc
+++ b/regression/python/nondet_str6/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.py
+--incremental-bmc
+^VERIFICATION FAILED$

--- a/regression/python/nondet_str7/main.py
+++ b/regression/python/nondet_str7/main.py
@@ -1,0 +1,46 @@
+# Test 1: Basic nondet_str() with reassignment on empty
+a = nondet_str()
+if a == "":
+    a = "default"
+
+# Test 2: Multiple independent nondet_str() allocations
+b = nondet_str()
+c = nondet_str()
+
+# Test 3: String reassignment
+d = nondet_str()
+d = "fixed"
+assert d == "fixed"
+
+# Test 4: Comparison with non-empty constant
+e = nondet_str()
+if e == "test":
+    e = "matched"
+else:
+    e = "other"
+
+# Test 5: List storage and retrieval (regression test)
+str1 = nondet_str()
+str2 = nondet_str()
+str3 = nondet_str()
+l = [str1, str2, str3]
+assert l[0] == str1
+assert l[1] == str2
+assert l[2] == str3
+
+# Test 6: Multiple reassignments
+f = nondet_str()
+f = "first"
+f = "second"
+assert f == "second"
+
+# Test 7: Empty string edge case
+g = nondet_str()
+if g == "":
+    g = "was_empty"
+    assert g == "was_empty"
+
+# Test 8: Inequality check
+h = nondet_str()
+i = nondet_str()
+result = (h == i)

--- a/regression/python/nondet_str7/test.desc
+++ b/regression/python/nondet_str7/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.py
+--unwind 20
+^VERIFICATION SUCCESSFUL$

--- a/regression/python/nondet_str8/main.py
+++ b/regression/python/nondet_str8/main.py
@@ -1,0 +1,4 @@
+x = nondet_str()
+# This assertion should FAIL because x is nondeterministic
+# It could be any string (including empty), not necessarily "hello"
+assert x == "hello"

--- a/regression/python/nondet_str8/test.desc
+++ b/regression/python/nondet_str8/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.py
+--incremental-bmc
+^VERIFICATION FAILED$

--- a/regression/python/nondet_str9/main.py
+++ b/regression/python/nondet_str9/main.py
@@ -1,0 +1,5 @@
+a = nondet_str()
+b = nondet_str()
+# This assertion should FAIL because a and b are independent nondeterministic values
+# They could be different strings
+assert a == b

--- a/regression/python/nondet_str9/test.desc
+++ b/regression/python/nondet_str9/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.py
+--incremental-bmc
+^VERIFICATION FAILED$

--- a/src/python-frontend/function_call_expr.cpp
+++ b/src/python-frontend/function_call_expr.cpp
@@ -252,16 +252,36 @@ exprt function_call_expr::build_nondet_call() const
     typet char_array_type =
       array_typet(char_type(), from_integer(max_str_length, size_type()));
 
-    exprt nondet_array = exprt("sideeffect", char_array_type);
-    nondet_array.statement("nondet");
+    // Create a temporary variable to hold the nondeterministic string
+    symbolt &nondet_str_symbol = converter_.create_tmp_symbol(
+      call_, "$nondet_str$", char_array_type, exprt());
 
+    // Declare the temporary
+    code_declt decl(symbol_expr(nondet_str_symbol));
+    decl.location() = converter_.get_location_from_decl(call_);
+    converter_.add_instruction(decl);
+
+    // Create nondet assignment for the array
+    exprt nondet_value("sideeffect", char_array_type);
+    nondet_value.statement("nondet");
+
+    code_assignt nondet_assign(symbol_expr(nondet_str_symbol), nondet_value);
+    nondet_assign.location() = converter_.get_location_from_decl(call_);
+    converter_.add_instruction(nondet_assign);
+
+    // Ensure null terminator at the last position
     exprt last_index = from_integer(max_str_length - 1, size_type());
     exprt null_char = from_integer(0, char_type());
 
-    with_exprt result(nondet_array, last_index, null_char);
-    result.type() = char_array_type;
+    index_exprt last_elem(symbol_expr(nondet_str_symbol), last_index);
+    code_assignt null_assign(last_elem, null_char);
+    null_assign.location() = converter_.get_location_from_decl(call_);
+    converter_.add_instruction(null_assign);
 
-    return result;
+    // Return address of first element: &arr[0] which is char*
+    index_exprt first_elem(
+      symbol_expr(nondet_str_symbol), from_integer(0, size_type()));
+    return address_of_exprt(first_elem);
   }
 
   exprt rhs = exprt("sideeffect", type_handler_.get_typet(type));

